File: 2024/Day8.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Day8 {
    public static void main(String[] args) throws IOException {
        BufferedReader f = new BufferedReader(new FileReader("day8.in"));

        int n = 12;
        char[][] grid = new char[n][n];
        for (int i = 0; i < n; i++) {
            String[] ss = f.readLine().split("");
            for (int j = 0; j < n; j++) {
                grid[i][j] = ss[j].charAt(0);
            }
        }

        f.close();

        // Part 1
        boolean[][] visited = new boolean[n][n];
        HashMap<Character, ArrayList<Pair>> freqs = new HashMap<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                char c = grid[i][j];

                if (c == '.')
                    continue;

                ArrayList<Pair> pl;
                if (freqs.containsKey(c)) {
                    pl = freqs.get(c);
                } else {
                    pl = new ArrayList<>();
                    freqs.put(c, pl);
                }
                pl.add(new Pair(i, j));
            }
        }

        for (Map.Entry<Character, ArrayList<Pair>> e : freqs.entrySet()) {
            ArrayList<Pair> fl = new ArrayList<>();
            for (int i = 0; i < fl.size(); i++) {
                for (int j = i; j < fl.size(); j++) {
                    // NEED TO FIGURE OUT HOW TO DETERMINE WHICH ONE IS ON "TOP"
                }
            }
        }
    }

    public static class Pair {
        public int x;
        public int y;

        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }

        @Override
        public String toString() {
            return "(" + x + ", " + y + ")";
        }

        @Override
        public boolean equals(Object o) {
            if (this == o)
                return true;
            if (o == null || getClass() != o.getClass())
                return false;
            Pair pair = (Pair) o;
            return x == pair.x && y == pair.y;
        }

        @Override
        public int hashCode() {
            return 31 * x + y;
        }
    }
}

File: 2024/Day8.py
f = open('day8.in')
g = list(map(lambda x: x.replace('\n', ''), f.readlines()))
n = len(g)

# Part 1
freqs = dict()
for i, w in enumerate(g):
    for j, u in enumerate(w):
        if u == '.':
            continue
        if u not in freqs:
            freqs[u] = set()
        freqs[u].add((i, j))

v = [[False]*n for i in range(n)]
c = 0
for k in freqs:
    fl = list(freqs[k])
    for i in range(len(fl)):
        a = fl[i]
        for j in range(i+1, len(fl)):
            b = fl[j]
            d = (a[0] - b[0], a[1] - b[1])
            ad = (a[0] + d[0], a[1] + d[1])
            bd = (b[0] - d[0], b[1] - d[1])
            if ad[0] >= 0 and ad[0] < n and ad[1] >= 0 and ad[1] < n and not v[ad[0]][ad[1]]:
                v[ad[0]][ad[1]] = True
                c += 1
            if bd[0] >= 0 and bd[0] < n and bd[1] >= 0 and bd[1] < n and not v[bd[0]][bd[1]]:
                v[bd[0]][bd[1]] = True
                c += 1
print(c)

# Part 2
v = [[False]*n for i in range(n)]
c = 0
freqs = dict()
for i, w in enumerate(g):
    for j, u in enumerate(w):
        if u == '.':
            continue
        if u not in freqs:
            freqs[u] = set()
        freqs[u].add((i, j))
        v[i][j] = True
        c += 1

for k in freqs:
    fl = list(freqs[k])
    for i in range(len(fl)):
        a = fl[i]
        for j in range(i+1, len(fl)):
            b = fl[j]
            d = (a[0] - b[0], a[1] - b[1])

            ad = (a[0] + d[0], a[1] + d[1])
            bd = (b[0] - d[0], b[1] - d[1])
            while ad[0] >= 0 and ad[0] < n and ad[1] >= 0 and ad[1] < n:
                if not v[ad[0]][ad[1]]:
                    v[ad[0]][ad[1]] = True
                    c += 1
                ad = (ad[0] + d[0], ad[1] + d[1])
            while bd[0] >= 0 and bd[0] < n and bd[1] >= 0 and bd[1] < n:
                if not v[bd[0]][bd[1]]:
                    v[bd[0]][bd[1]] = True
                    c += 1
                bd = (bd[0] - d[0], bd[1] - d[1])

print(c)